scipio/src/timer.rs

```diff
@@ -6,18 +6,47 @@
 use crate::parking::Reactor;
 use crate::task::JoinHandle;
 use crate::{Local, QueueNotFoundError, Task, TaskQueueHandle};
+use std::cell::RefCell;
 use std::future::Future;
 use std::pin::Pin;
+use std::rc::Rc;
 use std::task::{Context, Poll, Waker};
 use std::time::{Duration, Instant};
 
+#[derive(Debug)]
+struct Inner {
+    id: u64,
+
+    waker: Option<Waker>,
+
+    /// When this timer fires.
+    when: Instant,
+}
+
+impl Inner {
+    fn reset(&mut self, dur: Duration) {
+        if let Some(_) = self.waker.as_ref() {
+            // Deregister the timer from the reactor.
+            Reactor::get().remove_timer(self.id);
+        }
+
+        // Update the timeout.
+        self.when = Instant::now() + dur;
+
+        if let Some(waker) = self.waker.as_mut() {
+            // Re-register the timer with the new timeout.
+            Reactor::get().insert_timer(self.id, self.when, waker);
+        }
```

Comment:
Can you explain why we update `self.when` even if we don't insert a new timer? Also, in what circunstances `self.waker` can be `None`?

Reply:
I can, but notice please that this code is pre-existing, I am just moving here.

`self.waker` starts as None. It only gets assigned when the future is first polled() (call to `await`).

Imagine for instance that you schedule a function that will call `timer.await` for execution in a task queue that is very busy: it may take a long time for `waker` not to be `None`. There are no guarantees as to when that might happen.

We only need to go to the reactor on reset if the future was already polled and is pending (has a waker).
If the future was not polled, then all we have to do is update `when`: whenever it does get polled, it will be polled with the right `when`

In summary:
`when` has to be updated in all cases: 

* if the future was polled already, so we always have the most up2date information (waker will be `Some`, because we are waiting on it)
 * if the future was not polled, so we register the timer correctly for the right expiration (waker will be `None`)